sled-agent/src/storage/dump_setup.rs

```diff
@@ -13,46 +13,81 @@ use std::path::{Path, PathBuf};
 use std::sync::{Arc, Weak};
 use std::time::{Duration, SystemTime, SystemTimeError, UNIX_EPOCH};
 use tokio::sync::MutexGuard;
-
-pub struct DumpSetup {
-    worker: Arc<std::sync::Mutex<DumpSetupWorker>>,
-    _poller: std::thread::JoinHandle<()>,
-    log: Logger,
-}
-
-impl DumpSetup {
-    pub fn new(log: &Logger) -> Self {
-        let worker = Arc::new(std::sync::Mutex::new(DumpSetupWorker::new(
-            log.new(o!("component" => "DumpSetup-worker")),
-        )));
-        let worker_weak = Arc::downgrade(&worker);
-        let log_poll = log.new(o!("component" => "DumpSetup-archival"));
-        let _poller = std::thread::spawn(move || {
-            Self::poll_file_archival(worker_weak, log_poll)
-        });
-        let log = log.new(o!("component" => "DumpSetup"));
-        Self { worker, _poller, log }
-    }
-}
+use zone::Zone;
 
 // we sure are passing a lot of Utf8PathBufs around, let's be careful about it
 #[derive(
     AsRef, Clone, Debug, Deref, Eq, From, Hash, Ord, PartialEq, PartialOrd,
 )]
-struct DumpSlicePath(Utf8PathBuf);
+pub(self) struct DumpSlicePath(Utf8PathBuf);
 #[derive(
     AsRef, Clone, Debug, Deref, Eq, From, Hash, Ord, PartialEq, PartialOrd,
 )]
-struct DebugDataset(Utf8PathBuf);
+pub(self) struct DebugDataset(Utf8PathBuf);
 #[derive(
     AsRef, Clone, Debug, Deref, Eq, From, Hash, Ord, PartialEq, PartialOrd,
 )]
-struct CoreDataset(Utf8PathBuf);
+pub(self) struct CoreDataset(Utf8PathBuf);
+
+#[cfg(not(test))]
+mod zpool_wrapper {
+    use super::{CoreDataset, DebugDataset, GetMountpoint};
+    use derive_more::{Deref, From};
+    use illumos_utils::zpool::ZpoolName;
 
-#[derive(Deref)]
-struct CoreZpool(ZpoolName);
-#[derive(Deref)]
-struct DebugZpool(ZpoolName);
+    #[derive(Deref, From)]
+    pub(super) struct CoreZpool(pub ZpoolName);
+    #[derive(Deref, From)]
+    pub(super) struct DebugZpool(pub ZpoolName);
+
+    impl GetMountpoint for DebugZpool {
+        type NewType = DebugDataset;
+        const MOUNTPOINT: &'static str = sled_hardware::disk::DUMP_DATASET;
+    }
+    impl GetMountpoint for CoreZpool {
+        type NewType = CoreDataset;
+        const MOUNTPOINT: &'static str = sled_hardware::disk::CRASH_DATASET;
+    }
+}
+#[cfg(test)]
```

Comment:
See below for my comments about mockall - I think, where possible, we should prefer fakes over mocks.

To be more specific, for something like `DumpSetupWorker`:

## What we're doing
- When we call `update_disk_loadout`, I see that could trigger some interactions with the underlying host system (via commands to ZFS, coreadm, dumpadm, etc)

### With Mocks
- In the mock-based world, we need to find and intercept each of these calls, and inject the "mock" versions when conditionally compiling with `cfg(test)`
- For any nested testing, this *really* sucks -- if we want to write a test that depends on `dump_setup.rs`, it will need to "infer and create mock objects" to interact with the underlying host system

### With fakes
- In the fake-based world, we could create a structure that defines how we want to interact with the host OS system (how exactly we carve this out is a bit subjective), perhaps in the form of a trait.
- Creating the `DumpSetupWorker` could then take a `dyn trait` object as input, and use this interface to interact with the host.
- On a real system: this could be dead-simple, pass-through calls to illumos-specific calls
- On a test system: we could implement a version of this trait that emulates the host however we want